esp-hal-common/src/i2c.rs

```diff
@@ -0,0 +1,968 @@
+    }
+}
+
+/// I2C-specific transmission errors
+#[derive(Debug)]
+pub enum Error {
+    ExceedingFifo,
+    AckCheckFailed,
+    TimeOut,
+    ArbitrationLost,
+    ExecIncomplete,
+    CommandNrExceeded,
+}
+
+/// I2C-specific setup errors
+#[derive(Debug)]
+pub enum SetupError {
+    InvalidClkConfig,
+    PeripheralDisabled,
+}
+
+/// Pins used by the I2C interface
+///
+/// Note that any two pins may be used
+pub struct Pins<SDA: OutputPin + InputPin, SCL: OutputPin + InputPin> {
+    pub sda: SDA,
+    pub scl: SCL,
+}
+
+/// A generic I2C Command
+enum Command {
+    Start,
+    Stop,
+    Write {
+        /// This bit is to set an expected ACK value for the transmitter.
+        ack_exp: Ack,
+        /// Enables checking the ACK value received against the ack_exp value.
+        ack_check_en: bool,
+        /// Length of data (in bytes) to be written. The maximum length is 255,
+        /// while the minimum is 1.
+        length: u8,
+    },
+    Read {
+        /// Indicates whether the receiver will send an ACK after this byte has
+        /// been received.
+        ack_value: Ack,
+        /// Length of data (in bytes) to be read. The maximum length is 255,
+        /// while the minimum is 1.
+        length: u8,
+    },
+}
+
+impl From<Command> for u16 {
+    fn from(c: Command) -> u16 {
+        let opcode = match c {
+            Command::Start => Opcode::RSTART,
+            Command::Stop => Opcode::STOP,
+            Command::Write { .. } => Opcode::WRITE,
+            Command::Read { .. } => Opcode::READ,
+        };
+
+        let length = match c {
+            Command::Start | Command::Stop => 0,
+            Command::Write { length: l, .. } | Command::Read { length: l, .. } => l,
+        };
+
+        let ack_exp = match c {
+            Command::Start | Command::Stop | Command::Read { .. } => Ack::NACK,
+            Command::Write { ack_exp: exp, .. } => exp,
+        };
+
+        let ack_check_en = match c {
+            Command::Start | Command::Stop | Command::Read { .. } => false,
+            Command::Write {
+                ack_check_en: en, ..
+            } => en,
+        };
+
+        let ack_value = match c {
+            Command::Start | Command::Stop | Command::Write { .. } => Ack::NACK,
+            Command::Read { ack_value: ack, .. } => ack,
+        };
+
+        let mut cmd: u16 = length.into();
+
+        if ack_check_en {
+            cmd |= 1 << 8;
+        } else {
+            cmd &= !(1 << 8);
+        }
+
+        if ack_exp == Ack::NACK {
+            cmd |= 1 << 9;
+        } else {
+            cmd &= !(1 << 9);
+        }
+
+        if ack_value == Ack::NACK {
+            cmd |= 1 << 10;
+        } else {
+            cmd &= !(1 << 10);
+        }
+
+        cmd |= (opcode as u16) << 11;
+
+        cmd
+    }
+}
+
+enum OperationType {
+    WRITE = 0,
+    READ  = 1,
+}
+
+#[derive(Eq, PartialEq, Copy, Clone)]
+enum Ack {
+    ACK,
+    NACK,
+}
+
+#[allow(dead_code)]
```

Comment:
`END` is not currently used; will it ever be? I'd say let's just remove it, unless there's a reason to keep it.